### PR TITLE
Porting eBook - more scenarios

### PR DESCRIPTION
**Pull request by ardalis**
## Summary

This PR extends the migration scenarios covered in chapter 4 and adds many additional scenarios covering more advanced or niche topics.

### CODE REVIEW

**Comment by ardalis** [comment]
@nishanil  @IEvangelist

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 1]
```diff
@@ -112,10 +112,414 @@ public IActionResult Index()
...
+
+        Assert.Equal(5, model.CatalogItem.Id);
+        Assert.Equal("Roslyn Red Sheet", model.CatalogItem.Name);
+    }
+}
+```
+
+If the app being migrated has no integration tests, the migration process can be a great opportunity to add some. These tests can verify that the migrated app behaves as the team expects. When such tests are in place early in a migration, they can ensure that later migration efforts do not break previously migrated portions of the app. Given how easy it is to set up and run integration tests in ASP.NET Core, the return on the investment spent setting up such tests is usually pretty high.
+
+## WCF Client Configuration
```
Comment (IEvangelist) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:493]:
```suggestion
## WCF client configuration
```

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 2]
```diff
@@ -112,10 +112,414 @@ public IActionResult Index()
...
+
+If your app currently relies on WCF services as a client, this scenario is supported. However, you will need to [migrate your configuration](/aspnet/core/migration/configuration) from *web.config* to use the new *appsettings.json* file. Another option is to add any necessary configuration to your clients programmatically when you create them. For example:
+
+```csharp
+var wcfClient = new OrderServiceClient(
+            new BasicHttpBinding(BasicHttpSecurityMode.None),
+            new EndpointAddress("http://localhost:5050/OrderService.svc"));
```
Comment (IEvangelist) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:500]:
```suggestion
var wcfClient = new OrderServiceClient(
    new BasicHttpBinding(BasicHttpSecurityMode.None),
    new EndpointAddress("http://localhost:5050/OrderService.svc"));
```

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 3]
```diff
@@ -112,10 +112,414 @@ public IActionResult Index()
...
+```
+
+If your organization has extensive services built using WCF that your app relies on, consider migrating them to use gRPC instead. For more details on gRPC, why you may wish to migrate, and a detailed migration guide, consult the [gRPC for WCF Developers](/dotnet/architecture/grpc-for-wcf-developers/) ebook.
```
Comment (IEvangelist) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:503]:
```suggestion
If your organization has extensive services built using WCF that your app relies on, consider migrating them to use gRPC instead. For more details on gRPC, why you may wish to migrate, and a detailed migration guide, consult the [gRPC for WCF Developers](/dotnet/architecture/grpc-for-wcf-developers/) eBook.
```

**Review by IEvangelist — CHANGES_REQUESTED** [review]
Review complete: I really like these updates - thank you @ardalis

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 4]
```diff
@@ -112,10 +112,414 @@ public IActionResult Index()
...
+        config.EnableCors();
+        // ...
+    }
+}
+```
+
+Once this is added, you can configure allowed origins, headers, and methods using the `EnableCors` attribute, like so:
+
+```csharp
+[EnableCors(origins: "https://ardalis.com", headers: "*", methods: "*")]
```
[the remark this answers is not in the document]
Comment (jamesmontemagno) [reply docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:364]:
```suggestion
[EnableCors(origins: "https://dot.net", headers: "*", methods: "*")]
```

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 5]
```diff
@@ -112,10 +112,415 @@ public IActionResult Index()
 
 This will default to returning the data in JSON format. XML and other formats will be used [if the app has been configured with the appropriate formatter](/aspnet/core/web-api/advanced/formatting).
 
+## Custom model binding
+
+Most ASP.NET MVC and Web API apps make use of model binding. The default model binding syntax migrates fairly seamlessly between these apps and ASP.NET Core MVC. However, in some cases customers have written [custom model binders](/aspnet/web-api/overview/formats-and-model-binding/parameter-binding-in-aspnet-web-api#model-binders) to support specific model types or usage scenarios. Custom model binders in ASP.NET MVC and Web API projects use separate `IModelBinder` interfaces defined in `System.Mvc` and `System.Web.Http` namespaces, respectively. In both cases, the custom binder exposes a `Bind` method that accepts a controller or action context and a model binding context as arguments.
```
Comment (jamesmontemagno) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:117]:
```suggestion
Most ASP.NET MVC and Web API apps make use of model binding. The default model binding syntax migrates fairly seamlessly between these apps and ASP.NET Core MVC. However, in some cases customers have written [custom model binders](/aspnet/web-api/overview/formats-and-model-binding/parameter-binding-in-aspnet-web-api#model-binders) to support specific model types or usage scenarios. Custom model binders in ASP.NET MVC and Web API projects use separate `IModelBinder` interfaces defined in `System.Web.Mvc` and `System.Web.Http` namespaces, respectively. In both cases, the custom binder exposes a `Bind` method that accepts a controller or action context and a model binding context as arguments.
```

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 6]
```diff
@@ -112,10 +112,415 @@ public IActionResult Index()
...
+```csharp
+public static void ConfigureApis(HttpConfiguration config)
+{
+    config.Formatters.Add(new ProductCsvFormatter()); 
+}
+```
+
+In ASP.NET Core, the process is similar. ASP.NET Core supports both input formatters (used by model binding) and output formatters (used to format responses). Adding a custom formatter to output responses in a specific way involves inheriting from an appropriate base class and adding the formatter to MVC in `Startup`:
+
+```
```
Comment (Youssef1313) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:183]:
~~~suggestion
```csharp
~~~

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 7]
```diff
@@ -112,10 +112,415 @@ public IActionResult Index()
...
+```csharp
+[RouteArea("Admin")]
+public class SomeController : Controller
+```
+
+When using Areas, additional arguments are passed into HTML helper methods to generate links to actions in different areas:
+
+```html
+@Html.ActionLink("News", "Index", "News", new { area = "News" }, null)
+```
```
Comment (Youssef1313) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:401]: Doesn't seem the suitable language identifier

Comment (ardalis) [reply docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:401]: Is there a better one for Razor?

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 8]
```diff
@@ -112,10 +112,414 @@ public IActionResult Index()
 
 This will default to returning the data in JSON format. XML and other formats will be used [if the app has been configured with the appropriate formatter](/aspnet/core/web-api/advanced/formatting).
 
+## Custom model binding
+
+Most ASP.NET MVC and Web API apps make use of model binding. The default model binding syntax migrates fairly seamlessly between these apps and ASP.NET Core MVC. However, in some cases customers have written [custom model binders](/aspnet/web-api/overview/formats-and-model-binding/parameter-binding-in-aspnet-web-api#model-binders) to support specific model types or usage scenarios. Custom model binders in ASP.NET MVC and Web API projects use separate `IModelBinder` interfaces defined in `System.Mvc` and `System.Web.Http` namespaces, respectively. In both cases, the custom binder exposes a `Bind` method that accepts a controller or action context and a model binding context as arguments.
```
[the remark this answers is not in the document]
Comment (ardalis) [reply docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:117]: Should have been `System.Web.Mvc` - fixed.

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md (hunk 7 above)
Comment (Youssef1313) [reply docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:401]: I think cshtml may work well.

Comment (ardalis) [reply docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:401]: Ok, thanks. Found this: https://docs.microsoft.com/en-us/contribute/code-in-docs#supported-languages and agree. Will fix.

**Comment by ardalis** [comment]
:shipit: ?

File: docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md [hunk 9]
```diff
@@ -112,10 +112,415 @@ public IActionResult Index()
...
+```
+
+You'll find a complete list of base classes in the [Microsoft.AspNetCore.Mvc.Formatters](https://docs.microsoft.com/dotnet/api/microsoft.aspnetcore.mvc.formatters) namespace.
+
+The steps to migrate from a Web API formatter to an ASP.NET Core MVC formatter are:
+
+1. Identify an appropriate base class for the new formatter.
+2. Create a new instance of the base class and implement its required methods.
+3. Copy over the functionality from the Web API formatter to the new implementation.
+4. Configure MVC in the ASP.NET Core App's `ConfigureServices` method to use the new formatter.
```
Comment (IEvangelist) [inline docs/architecture/porting-existing-aspnet-apps/more-migration-scenarios.md:201]:
```suggestion
1. Identify an appropriate base class for the new formatter.
1. Create a new instance of the base class and implement its required methods.
1. Copy over the functionality from the Web API formatter to the new implementation.
1. Configure MVC in the ASP.NET Core App's `ConfigureServices` method to use the new formatter.
```

**Review by IEvangelist — APPROVED** [review]

**Pull request merged**